src/passes/Heap2Local.cpp

```diff
@@ -518,11 +523,10 @@ struct Heap2LocalOptimizer {
     Mixes,
   };
 
-  // Analyze an allocation to see if we can convert it from a heap allocation to
-  // locals.
-  void convertToLocals(StructNew* allocation) {
-    Rewriter rewriter(allocation, func, module);
-
+  // Analyze an allocation to see if it escapes or not. We receive a Rewriter
+  // instance on which we store important information as we go, which will be
+  // necessary if we optimize later.
+  bool escapes(StructNew* allocation, Rewriter& rewriter) {
```

Comment:
It's odd that the analysis depends on the rewriter. Would it be much work to factor out the parts of the rewriter that the analysis uses as well?

Reply:
It's more a matter of efficiency. To do this in a single pass versus two passes, it makes sense to check for escaping while we note what we need to fix up if we do find nothing escapes.

Reply:
Right, I don't mean that we should do this in more passes, but it would be nice to move `rewriter.sets` and `rewriter.reaches` out of the rewriter and into the analyzer instead.

Reply:
Hmm, interesting. Yeah, maybe a larger refactoring is worthwhile here. I'll look into that and update this PR with my findings.